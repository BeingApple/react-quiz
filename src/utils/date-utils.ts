import moment from "moment"
import { Moment, unitOfTime } from "moment"

type Unit = {
  unit: unitOfTime.Base
  string: string
}

const units: Array<Unit> = [
  {unit: "years", string: "년"},
  {unit: "months", string: "개월"}, 
  {unit: "days", string: "일"}, 
  {unit: "hours", string: "시간"}, 
  {unit: "minutes", string: "분"},
  {unit: "seconds", string: "초"},
]

export const durationToString = (duration: moment.Duration) => {
  let string = ""

  for (const unit of units) {
    const unitTime = duration.get(unit.unit)
    if (unitTime > 0) {
      string += unitTime + unit.string + " "
    }
  }

  return string.trim();
}

export const dateToString = (date?: Moment) => {
  date = (typeof date === 'string' || date instanceof String) ? moment(date) : date

  return date?.format("yyyy-MM-DD HH:mm:ss") ?? ""
}