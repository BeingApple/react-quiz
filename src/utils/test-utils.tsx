import React, { PropsWithChildren } from 'react'
import { render } from '@testing-library/react'
import type { RenderOptions } from '@testing-library/react'
import { Provider as ReduxProvider } from 'react-redux'

import { setupStore, type AppStore, type RootState } from '@/store/store'
// As a basic setup, import your same slice reducers
import { QueryClient, QueryClientProvider } from '@tanstack/react-query'

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: Partial<RootState>
  store?: AppStore
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  const queryClient = new QueryClient()

  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return (
      <ReduxProvider store={store}>
        <QueryClientProvider client={queryClient}>
          {children}
        </QueryClientProvider>
      </ReduxProvider>
    )
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}