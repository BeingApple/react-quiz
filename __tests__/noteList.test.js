import '@testing-library/jest-dom'
import NoteList from '@/components/domain/note/list'
import { renderWithProviders } from '../src/utils/test-utils'
import moment from 'moment';
import { dateToString } from "@/utils/date-utils";

describe('<NoteList />', () => {
  const sampleItem = {
    type: 'multiple',
    difficulty: 'medium',
    category: 'TEST CATEGORY',
    question: 'TEST QUESTION',
    correct_answer: 'CORRENT ANSWER',
    incorrect_answers: ['INCORRECT ANSWER 1', 'INCORRECT ANSWER 2', 'INCORRECT ANSWER 3'],
    answers: ['CORRENT ANSWER', 'INCORRECT ANSWER 1', 'INCORRECT ANSWER 2', 'INCORRECT ANSWER 3'],
    select_answer: 'INCORRECT ANSWER 2'
  }
  const sampleNote = {
    recordAt: moment('2024-02-19 00:00:00'),
    wrongList: [sampleItem, sampleItem, sampleItem]
  }

  const setup = () => {
    const initialState = { note: {
      noteList: [sampleNote]
    }};
    const utils = renderWithProviders(<NoteList/>, {preloadedState: initialState})
    const { getByText } = utils

    const item = initialState.note.noteList[0]
    const titleString = `${dateToString(item.recordAt)}의 퀴즈 풀이`

    const listTitle = getByText(titleString)

    return {
      ...utils,
      listTitle,
    };
  };

  it('matches snapshot', () => {
    const { container } = setup()
    expect(container).toMatchSnapshot();
  });

  it('has list', () => {
    const { listTitle } = setup()
    expect(listTitle).toBeTruthy()
  });
});