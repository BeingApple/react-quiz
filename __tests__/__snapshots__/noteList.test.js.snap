// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NoteList /> matches snapshot 1`] = `
<div>
  <div
    class="MuiBox-root css-144ebsl"
  >
    <ul
      class="MuiList-root MuiList-padding css-h4y409-MuiList-root"
    >
      <li
        class="MuiListSubheader-root MuiListSubheader-gutters MuiListSubheader-sticky css-uob957-MuiListSubheader-root"
      >
        오답노트
      </li>
      <li
        class="MuiListItem-root MuiListItem-gutters css-1nxmd3h-MuiListItem-root"
      >
        <div
          class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
          role="button"
          tabindex="0"
        >
          <div
            class="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="RuleIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.54 11 13 7.46l1.41-1.41 2.12 2.12 4.24-4.24 1.41 1.41zM11 7H2v2h9zm10 6.41L19.59 12 17 14.59 14.41 12 13 13.41 15.59 16 13 18.59 14.41 20 17 17.41 19.59 20 21 18.59 18.41 16zM11 15H2v2h9z"
              />
            </svg>
          </div>
          <div
            class="MuiListItemText-root css-tlelie-MuiListItemText-root"
          >
            <span
              class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
            >
              2024-02-19 00:00:00의 퀴즈 풀이
            </span>
          </div>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </div>
      </li>
    </ul>
  </div>
</div>
`;
